### ISSUE TITLE
Support Provided Union and Record Types

### ISSUE THREAD

**Issue by baronfel** [baronfel/fsharp-lang-suggestions-final#154]
# Support Provided Union and Record Types [5663267] #

**Submitted by Gustavo Guerra on 3/21/2014 12:00:00 AM**  
**111 votes on UserVoice prior to migration**  

## Response ##
** by fslang-admin on 2/4/2016 12:00:00 AM **

Marking as “approved in principle” (it will show as “planned”).
However, in reality it is both relatively low priority and relatively hard to implement, so I don’t expect this to happen any time soon.
For example, we would have to reverse-map the metadata associated with F#-specific types. This is already done in F# reflection FSharpType.IsRecord and so on.

**[Original UserVoice Submission](https://fslang.uservoice.com/forums/245727-f-language/suggestions/5663267)**
[Archived Uservoice Comments](../tree/master/archive/suggestion-5663267-support-provided-union-and-record-types.md#comments)

**Comment by dsyme** [comment]
** by fslang-admin on 2/4/2016 12:00:00 AM **

Marking as “approved in principle” (it will show as “planned”). However, in reality it is both relatively low priority and relatively hard to implement, so I don’t expect this to happen any time soon. For example, we would have to reverse-map the metadata associated with F#-specific types. This is already done in F# reflection FSharpType.IsRecord and so on.

**Comment by 7sharp9** [comment]
I propose we split this in two, add provided records, add provided unions.  

@dsyme I know you said provided active patterns could be more generally usable than unions, maybe this could be fleshed out?

**Comment by dsyme** [comment]
> I propose we split this in two, add provided records, add provided unions.

For RFCs yes, we can do two separate ones

> @dsyme I know you said provided active patterns could be more generally usable than unions, maybe this could be fleshed out?

* APs are defined as module-let functions. Provided module-let functions and values are not supported.  Hence the first step would probably be to allow those, then APs should drop out as part of that.

**Comment by 7sharp9** [comment]
@dsyme I actually have ProvidedUnions and ProvidedRecords in Falanx :-)
_obviously I have different parameters and constraints_

**Comment by dsyme** [comment]
Added a relevant comment to one of the above discussions:

> 1.	Why isn't it possible to generate F# specific types (aren't they just nested classes with custom attributes/metadata after all) ?

The main thing is that it's just a lot of work.  The way TPs are architected (providing .NET System.Type/MethodInfo etc. and the compiler converts them to F# TAST nodes) effectively means we have to fully invert the compilation process from .NET IL types --> F# TAST types.   This is implemented in tast.fs and infos.fs in the F# compiler for methods, types, properties, events but there is no corresponding inversion for F#-defined union/record types.

The inversion would be similar to that implemented in FSharp.Reflection.FSharpType/FSharpValue based on the attributes etc.   But it's painful to implement and fully test accurately.
 

 
> 2.	Is there a general interest in making this possible (or is it just me wanting this) ?

Yes, people are interested in this

**Comment by 7sharp9** [comment]
Yeah I bake in all the attributes to get reflection to work for ProvidedUnions and ProvidedRecords in Falanx.

**Comment by realvictorprm** [comment]
I just noticed, we don't have an RFC for this.

I'll dig into some stuff and see whether I can get anything together. I have the feeling this is worth having for FS-1023

**Comment by Luiz-Monad** [comment]
> The main thing is that it's just a lot of work. The way TPs are architected (providing .NET System.Type/MethodInfo etc. and the compiler converts them to F# TAST nodes) effectively means we have to fully invert the compilation process from .NET IL types --> F# TAST types. This is implemented in tast.fs and infos.fs in the F# compiler for methods, types, properties, events but there is no corresponding inversion for F#-defined union/record types.

Wouldn't it be interesting to allow TypeProviders to create something closer to the TAST types directly instead of relying on System.Type ?
